module/VuFind/src/VuFind/Db/Service/RecordService.php

```diff
@@ -0,0 +1,140 @@
+ * @link     https://vufind.org/wiki/development:plugins:database_gateways Wiki
+ */
+
+namespace VuFind\Db\Service;
+
+use Exception;
+use VuFind\Db\Entity\RecordEntityInterface;
+use VuFind\Db\Table\DbTableAwareInterface;
+use VuFind\Db\Table\DbTableAwareTrait;
+
+/**
+ * Database service for Records.
+ *
+ * @category VuFind
+ * @package  Database
+ * @author   Sudharma Kellampalli <skellamp@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:database_gateways Wiki
+ */
+class RecordService extends AbstractDbService implements DbTableAwareInterface, RecordServiceInterface
+{
+    use DbTableAwareTrait;
+
+    /**
+     * Find a record by id.
+     *
+     * @param string $id     Record ID
+     * @param string $source Record source
+     *
+     * @return ?RecordEntityInterface
+     */
+    public function findRecord($id, $source): ?RecordEntityInterface
+    {
+        return $this->getDbTable('record')->findRecord($id, $source);
+    }
+
+    /**
+     * Find records by ids.
+     *
+     * @param array  $ids    Record IDs
+     * @param string $source Record source
+     *
+     * @return RecordEntityInterface[] Array of record objects found
+     */
+    public function findRecords($ids, $source): array
+    {
+        return $this->getDbTable('record')->findRecords($ids, $source);
+    }
+
+    /**
+     * Update an existing entry in the record table or create a new one.
+     *
+     * @param string $id      Record ID
+     * @param string $source  Data source
+     * @param string $rawData Raw data from source
+     *
+     * @return RecordEntityInterface
+     */
+    public function updateRecord($id, $source, $rawData): RecordEntityInterface
+    {
+        $record = $this->findRecord($id, $source);
+        if (!$record) {
+            $record = $this->createEntity();
+        }
+        $record->setRecordId($id)
+            ->setSource($source)
+            ->setData(serialize($rawData))
+            ->setVersion(\VuFind\Config\Version::getBuildVersion())
+            ->setUpdated(new \DateTime());
+        $this->persistEntity($record);
+        return $record;
+    }
+
+    /**
+     * Clean up orphaned entries (i.e. entries that are not in favorites anymore)
+     *
+     * @return int Number of records deleted
+     */
+    public function cleanup(): int
+    {
+        return $this->getDbTable('record')->cleanup();
+    }
+
+    /**
+     * Delete a record by source and id. Return true if found and deleted, false if not found.
+     * Throws exception if something goes wrong.
+     *
+     * @param string $id     Record ID
+     * @param string $source Record source
+     *
+     * @return bool
+     * @throws Exception
+     */
+    public function deleteRecord(string $id, string $source): bool
+    {
+        $record = $this->findRecord($id, $source);
```

Comment:
Would it be more correct to use `$record = this->getDbTable('record')->findRecord($id, $source);`?

Reply:
Again, I was trying to reduce diffs with #2233, but your proposal is more technically correct, so I have adopted it.